opm/input/eclipse/Schedule/Well/Well.cpp

```diff
@@ -1733,10 +1733,14 @@ int Well::vfp_table_number() const {
 }
 
 double Well::temperature() const {
-    if (!this->wtype.producer())
-        return this->well_temperature;
+    if (!this->wtype.producer() && this->well_temperature)
+        return *this->well_temperature;
 
-    throw std::runtime_error("Can only ask for temperature in an injector");
+    throw std::runtime_error("Can only ask for well temperature for" 
+                    "injectors with non-default temperature.");
+}
+bool Well::hasTemperature() const {
+    return this->well_temperature.has_value(); 
```

Comment:
If the assumption that the temperature is only related to injectors stands, the function names `hasTemperature` and `tempature` can also change to reflect that the temperature is for injectors only. (there is a dedicated function isInjector() defined, which can be better for readability?) 

And we can check production/injection types in this function. 

And the if condition above `if (!this->wtype.producer() && this->well_temperature)` can become 
`if (this->hasTemperature) or `if (this->hasInjTemperature)` if the function name changes.